File: Base/reg_stats.py
"""Register Allocation Statistics Helper

* FunRegStats needs to be accurate because it is used to make spilling decisions
* The terms global and local register are WRT to the function
* A registers is NOT global just because it is referenced in multiple Bbls
  it actually has to be live across Bbls (appear in a live-out set)
* A Bbl may contain mutliple calls
* "lac" stands for "live across calls", both local and global registers may be "lac"

"""

import collections
import dataclasses
from typing import List, Dict, Tuple, Set

from Base import ir
from Base import liveness
from Base import reg_alloc
import Base.opcode_tab as o

REG_KIND_LAC = Tuple[o.DK, bool]


class BblRegUsageStatsRegPool(reg_alloc.RegPool):
    """Regpool for determining register pressure at the Bbl level

    Can be used repeatedly in which case it will  compute the maximum
    register pressure over a set of Bbls
    """

    def __init__(self, reg_kind_map: Dict[o.DK, o.DK]):
        super(BblRegUsageStatsRegPool, self).__init__()
        self.counter = 0
        self.reg_kind_map = reg_kind_map
        self._available: Dict[REG_KIND_LAC,
                              List[ir.CpuReg]] = collections.defaultdict(list)

    def _get_reg_class(self, lr: reg_alloc.LiveRange) -> REG_KIND_LAC:
        return self.reg_kind_map[lr.reg.kind], liveness.LiveRangeFlag.LAC in lr.flags

    # @override
    def get_available_reg(self, lr: liveness.LiveRange) -> ir.CpuReg:
        key: REG_KIND_LAC = self._get_reg_class(lr)
        available = self._available.get(key)
        if available:
            return available.pop(-1)
        else:
            # manufacture a new register
            self.counter += 1
            return ir.CpuReg(f"z{self.counter}", key[1], key[0].value)

    def usage(self) -> Dict[REG_KIND_LAC, int]:
        assert sum(len(val)
                   for val in self._available.values()) == self.counter
        return {key: len(val) for (key, val) in self._available.items()}

    # @override
    def give_back_available_reg(self, cpu_reg: ir.CpuReg):
        key: REG_KIND_LAC = (o.DK(cpu_reg.kind), bool(cpu_reg.no))
        self._available[key].append(cpu_reg)


def LiveRangeShouldBeIgnored(lr: liveness.LiveRange, reg_kind_map) -> bool:
    if lr.is_cross_bbl(): return True
    if lr.is_use_lr(): return False  # this is an entry for "uses"
    if lr.reg.kind not in reg_kind_map: return True
    return lr.reg.HasCpuReg()


def FunComputeBblRegUsageStats(fun: ir.Fun,
                               reg_kind_map: Dict[o.DK, o.DK]) -> Dict[REG_KIND_LAC, int]:
    """
    Computes maximum number of register needed for locals across all Bbls

    Requires liveness
    """
    pool = BblRegUsageStatsRegPool(reg_kind_map)
    for bbl in fun.bbls:
        live_ranges = liveness.BblGetLiveRanges(bbl, fun, bbl.live_out, True)
        # we do not want re-use of regs that are not coming from the pool
        for lr in live_ranges:
            if LiveRangeShouldBeIgnored(lr, reg_kind_map):
                lr.flags |= liveness.LiveRangeFlag.IGNORE
        reg_alloc.RegisterAssignerLinearScan(live_ranges, pool)
    return pool.usage()


def FunComputeRegStatsExceptLAC(fun: ir.Fun):
    """Updates Reg info: Sets def_ins, def_bbl and the flags:

    GLOBAL, MULTIDEF, IS_READ. MULTI_READ

    Note, the computation of GLOBAL is conservative in that locally used
    regs may appear global if they occur in multiple bbls.

    Once we have renamed those registers via FunSeparateLocalRegUsage the issue goes away
    """
    for reg in fun.regs:
        reg.def_ins = ir.INS_INVALID
        reg.def_bbl = ir.BBL_INVALID
        reg.flags = ir.REG_FLAG(0)
    for bbl in fun.bbls:
        for ins in bbl.inss:
            num_defs = ins.opcode.def_ops_count()
            for n, reg in enumerate(ins.operands):
                if not isinstance(reg, ir.Reg):
                    continue
                if n < num_defs:
                    if reg.def_ins is ir.INS_INVALID:
                        reg.def_ins = ins
                        reg.def_bbl = bbl
                    else:
                        reg.flags |= ir.REG_FLAG.MULTI_DEF
                        if bbl != reg.def_bbl:
                            reg.flags |= ir.REG_FLAG.GLOBAL
                else:
                    if reg.flags & ir.REG_FLAG.IS_READ:
                        reg.flags |= ir.REG_FLAG.MULTI_READ
                    else:
                        reg.flags |= ir.REG_FLAG.IS_READ
                    if bbl != reg.def_bbl:
                        reg.flags |= ir.REG_FLAG.GLOBAL


def FunComputeRegStatsLAC(fun: ir.Fun):
    """Updates Reg info: Sets flags: GLOBAL, LAC

    Note the GLOBAL flags computation is more accurate than FunComputeRegStatsExceptLAC.
    """
    for reg in fun.regs:
        reg.flags &= ~(ir.REG_FLAG.GLOBAL | ir.REG_FLAG.LAC)
    for bbl in fun.bbls:
        live_out = bbl.live_out.copy()
        for ins in reversed(bbl.inss):
            if ins.opcode.is_call():
                for reg in live_out:
                    reg.flags |= ir.REG_FLAG.LAC
            num_defs = ins.opcode.def_ops_count()
            for n, reg in enumerate(ins.operands):
                if not isinstance(reg, ir.Reg): continue
                if n < num_defs:
                    live_out.discard(reg)
                else:
                    live_out.add(reg)
        for reg in live_out:
            reg.flags |= ir.REG_FLAG.GLOBAL


def FunDropUnreferencedRegs(fun: ir.Fun) -> int:
    """Remove all regs which are no longer referenced"""
    to_be_removed: List[ir.Reg] = []
    for reg in fun.regs:
        if reg.def_ins == ir.INS_INVALID and ir.REG_FLAG.IS_READ not in reg.flags:
            to_be_removed.append(reg)
    for reg in to_be_removed:
        fun.regs.remove(reg)
        del fun.reg_syms[reg.name]
    return len(to_be_removed)


@dataclasses.dataclass(init=True)
class FunRegStats:
    global_lac = 0
    global_not_lac = 0
    local_lac = 0
    local_not_lac = 0

    def __str__(self):
        return (
            f"{self.global_lac:2}/{self.global_not_lac:2}  "
            f"{self.local_lac:2}/{self.local_not_lac:2}")


def FunCalculateRegStats(fun: ir.Fun) -> FunRegStats:
    rs = FunRegStats()
    for reg in fun.regs:
        if ir.REG_FLAG.GLOBAL in reg.flags:
            if ir.REG_FLAG.LAC in reg.flags:
                rs.global_lac += 1
            else:
                rs.global_not_lac += 1
        else:
            if ir.REG_FLAG.LAC in reg.flags:
                rs.local_lac += 1
            else:
                rs.local_not_lac += 1
    return rs


def _BblRenameReg(bbl, pos, reg_src, reg_dst):
    done = False
    for ins in bbl.inss[pos:]:
        num_defs = ins.opcode.def_ops_count()
        for n, reg in enumerate(ins.operands):
            if reg != reg_src:
                continue
            if n < num_defs:
                done = True
            else:
                ins.operands[n] = reg_dst
        if done:
            return


def FunSeparateLocalRegUsage(fun: ir.Fun) -> int:
    count = 0
    for bbl in fun.bbls:
        for pos, ins in enumerate(bbl.inss):
            num_defs = ins.opcode.def_ops_count()
            for n, reg in enumerate(ins.operands[:num_defs]):
                assert isinstance(reg, ir.Reg)
                if reg.def_ins is ins or ir.REG_FLAG.GLOBAL in reg.flags or reg.cpu_reg is not None:
                    continue
                purpose = reg.name
                if purpose.startswith("$"):
                    underscore_pos = purpose.find("_")
                    purpose = purpose[underscore_pos + 1:]
                new_reg = fun.GetScratchReg(reg.kind, purpose, False)
                ins.operands[n] = new_reg
                _BblRenameReg(bbl, pos + 1, reg, new_reg)
                count += 1
    return count


def FunPreallocatedRegs(fun: ir.Fun) -> Set[ir.CpuReg]:
    out: Set[ir.CpuReg] = set()
    for reg in fun.regs:
        if reg.cpu_reg:
            out.add(reg.cpu_reg)
    return out

File: Base/reg_stats.cc
// (c) Robert Muth - see LICENSE for more info

#include "Base/reg_stats.h"
#include "Base/liveness.h"
#include "Base/opcode_gen.h"
#include "Base/reg_alloc.h"
#include "Base/serialize.h"

#include <iomanip>
#include <map>

// #define TRACE_REG_ALLOC

namespace cwerg::base {
namespace {

// +-prefix converts an enum the underlying type
template <typename T>
constexpr auto operator+(T e) noexcept
    -> std::enable_if_t<std::is_enum<T>::value, std::underlying_type_t<T>> {
  return static_cast<std::underlying_type_t<T>>(e);
}

class BblRegUsageStatsRegPool : public RegPool {
 public:
  explicit BblRegUsageStatsRegPool(const DK_MAP& rk_map) : rk_map_(rk_map) {}

  CpuReg get_available_reg(const LiveRange& lr) override {
    const DK dk = get_rk(lr.reg);
    const bool lac = lr.HasFlag(LR_FLAG::LAC);
    std::vector<CpuReg>& pool = (lac ? available_lac_ : available_not_lac_)[dk];
    // Note: the cpu_reg is a fake handle
    CpuReg cpu_reg;
    if (pool.empty()) {
      ++counter_;
      cpu_reg = CpuReg(counter_ | (+dk << 16U) | (lac << 15U));
    } else {
      cpu_reg = pool.back();
      pool.pop_back();
    }
#ifdef TRACE_REG_ALLOC
    std::cout << (cpu_reg.index() & 0xffff) << " " << int(dk) << " <- " << lr
              << "\n";
#endif
    return cpu_reg;
  }

  uint8_t get_cpu_reg_family(DK dk) override { return rk_map_[+dk]; }

  void give_back_available_reg(CpuReg cpu_reg) override {
#ifdef TRACE_REG_ALLOC
    std::cout << "FREE: " << (cpu_reg.index() & 0xffff) << "\n";
#endif
    const DK dk = static_cast<DK>((cpu_reg.index() >> 16U) & 0xff);
    const bool lac = (cpu_reg.index() >> 15U) & 1U;
    std::vector<CpuReg>& pool = (lac ? available_lac_ : available_not_lac_)[dk];
    pool.push_back(cpu_reg);
  }

  DK_LAC_COUNTS usage() const {
    DK_LAC_COUNTS out;
    for (auto& [k, v] : available_lac_) {
      out.lac[+k] += v.size();
    }
    for (auto& [k, v] : available_not_lac_) {
      out.not_lac[+k] += v.size();
    }
    unsigned count = 0;
    for (size_t i = 0; i < out.lac.size(); ++i) {
      count += out.lac[i] + out.not_lac[i];
    }
    ASSERT(count == counter_, "expected reg_num: " << counter_ << " actual "
                                                   << count << "\n"
                                                   << out);
    return out;
  }

  unsigned counter() const { return counter_; }

 private:
  DK get_rk(Reg reg) {
    const DK out = DK(rk_map_[+RegKind(reg)]);
    ASSERT(out != DK::INVALID, "reg has untracked type: " << Name(reg));
    return out;
  }

  const DK_MAP& rk_map_;

  std::map<DK, std::vector<CpuReg>> available_lac_;
  std::map<DK, std::vector<CpuReg>> available_not_lac_;
  unsigned counter_ = 0;
};

bool LiveRangeShouldBeIgnored(const LiveRange& lr, const DK_MAP& rk_map) {
  if (lr.is_cross_bbl()) return true;
  if (lr.is_use_lr()) return false;
  if (rk_map[+RegKind(lr.reg)] == +DK::INVALID) return true;
  return !RegCpuReg(lr.reg).isnull();
}

}  // namespace

DK_LAC_COUNTS FunComputeBblRegUsageStats(Fun fun, const DK_MAP& rk_map) {
  const unsigned num_regs = FunNumRegs(fun);
  const Reg* const reg_map = (Reg*)FunRegMap(fun).BackingStorage();
  std::vector<Reg> live_out;

  BblRegUsageStatsRegPool pool(rk_map);
  for (Bbl bbl : FunBblIter(fun)) {
    live_out.clear();
    const BitVec& live_out_vec = BblLiveOut(bbl);
    for (int i = 1; i < num_regs; ++i) {
      if (live_out_vec.BitGet(i)) live_out.push_back(reg_map[i]);
    }

    std::vector<LiveRange> ranges = BblGetLiveRanges(bbl, fun, live_out, true);
    std::vector<LiveRange*> ordered;
    // Skip the first range which is a dummy
    for (unsigned i = 1; i < ranges.size(); ++i) ordered.push_back(&ranges[i]);
    std::sort(begin(ordered), end(ordered),
              [](LiveRange* lhs, LiveRange* rhs) { return *lhs < *rhs; });
#ifdef TRACE_REG_ALLOC
    std::cout
        << "@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@\n";
    BblRenderToAsm(bbl, fun, &std::cout);
    for (const auto& lr : ordered) {
      std::cout << *lr << "\n";
    }
#endif
    for (LiveRange& lr : ranges) {
      if (LiveRangeShouldBeIgnored(lr, rk_map)) {
        lr.SetFlag(LR_FLAG::IGNORE);
      }
    }
    RegisterAssignerLinearScan(ordered, &ranges, &pool);
  }

  return pool.usage();
}

std::ostream& operator<<(std::ostream& os, const DK_LAC_COUNTS& usage) {
  unsigned lac = 0;
  unsigned not_lac = 0;
  for (size_t i = 0; i < usage.lac.size(); ++i) {
    lac += usage.lac[i];
    not_lac += usage.not_lac[i];
  }
  os << std::setw(2) << lac << "/" << std::setw(2) << not_lac;
  return os;
}

void FunComputeRegStatsExceptLAC(Fun fun) {
  for (Reg reg : FunRegIter(fun)) {
    RegFlags(reg) &=
        ~(+REG_FLAG::MULTI_DEF | +REG_FLAG::GLOBAL | +REG_FLAG::IS_READ);
    RegDefIns(reg) = Ins(0);
    RegDefBbl(reg) = Bbl(0);
  }
  for (Bbl bbl : FunBblIter(fun)) {
    for (Ins ins : BblInsIter(bbl)) {
      const unsigned num_defs = InsOpcode(ins).num_defs;
      const unsigned num_ops = InsOpcode(ins).num_operands;
      for (unsigned i = 0; i < num_defs; ++i) {
        Reg reg = Reg(InsOperand(ins, i));  // must be a reg
        if (RegDefIns(reg).isnull()) {      // first definition
          RegDefIns(reg) = ins;
          RegDefBbl(reg) = bbl;
        } else {  // not the first definition
          RegFlags(reg) |= +REG_FLAG::MULTI_DEF;
          if (RegDefBbl(reg) != bbl) {
            RegFlags(reg) |= +REG_FLAG::GLOBAL;
          }
        }
      }
      for (unsigned i = num_defs; i < num_ops; ++i) {
        Reg reg = Reg(InsOperand(ins, i));
        if (reg.kind() != RefKind::REG) continue;
        if (RegHasFlag(reg, REG_FLAG::IS_READ)) {
          RegFlags(reg) |= +REG_FLAG::MULTI_READ;
        } else {
          RegFlags(reg) |= +REG_FLAG::IS_READ;
        }
        if (RegDefBbl(reg) != bbl) {
          RegFlags(reg) |= +REG_FLAG::GLOBAL;
        }
      }
    }
  }
}

void FunComputeRegStatsLAC(Fun fun) {
  for (Reg reg : FunRegIter(fun)) {
    // we will (re)compute these
    RegFlags(reg) &= ~(+REG_FLAG::GLOBAL | +REG_FLAG::LAC);
  }
  const unsigned num_regs = FunNumRegs(fun);
  BitVec live = BitVec::New(num_regs);
  const Reg* const reg_map = (Reg*)FunRegMap(fun).BackingStorage();
  for (Bbl bbl : FunBblIter(fun)) {
    live.CopyFrom(BblLiveOut(bbl));
    for (Ins ins : BblInsIterReverse(bbl)) {
      if (InsOpcode(ins).IsCall()) {
        for (unsigned i = 1; i < num_regs; ++i) {
          if (live.BitGet(i)) {
            RegFlags(reg_map[i]) |= +REG_FLAG::LAC;
          }
        }
      }
      const unsigned num_defs = InsOpcode(ins).num_defs;
      const unsigned num_ops = InsOpcode(ins).num_operands;
      for (unsigned i = 0; i < num_defs; ++i) {
        const Reg reg = Reg(InsOperand(ins, i));
        live.BitClear(RegNo(reg));
      }

      for (unsigned i = num_defs; i < num_ops; ++i) {
        const Reg reg = Reg(InsOperand(ins, i));
        if (reg.kind() != RefKind::REG) continue;
        live.BitSet(RegNo(reg));
      }
    }
    for (unsigned i = 1; i < num_regs; ++i) {
      if (live.BitGet(i)) {
        RegFlags(reg_map[i]) |= +REG_FLAG::GLOBAL;
      }
    }
  }
  BitVec::Del(live);
}

int FunDropUnreferencedRegs(Fun fun) {
  std::vector<Reg> to_be_deleted;
  for (Reg reg : FunRegIter(fun)) {
    if (!RegHasFlag(reg, REG_FLAG::IS_READ) && RegDefIns(reg).isnull()) {
      to_be_deleted.push_back(reg);
    }
  }
  for (Reg reg : to_be_deleted) {
    FunRegDel(fun, reg);
    RegDel(reg);
  }

  return to_be_deleted.size();
}

FunRegStats FunCalculateRegStats(Fun fun) {
  FunRegStats rs;
  for (Reg reg : FunRegIter(fun)) {
    if (RegHasFlag(reg, REG_FLAG::GLOBAL)) {
      if (RegHasFlag(reg, REG_FLAG::LAC)) {
        ++rs.global_lac;
      } else {
        ++rs.global_not_lac;
      }
    } else {
      if (RegHasFlag(reg, REG_FLAG::LAC)) {
        ++rs.local_lac;
      } else {
        ++rs.local_not_lac;
      }
    }
  }
  return rs;
}

std::ostream& operator<<(std::ostream& os, const FunRegStats& stats) {
  os << std::right << std::setw(2) << stats.global_lac << "/" << std::setw(2)
     << stats.global_not_lac << "  " << std::setw(2) << stats.local_lac << "/"
     << std::setw(2) << stats.local_not_lac;

  return os;
}

namespace {
void RenameRegRange(Ins ins, Reg reg_src, Reg reg_dst) {
  while (!BblInsList::IsSentinel(ins)) {
    const unsigned num_defs = InsOpcode(ins).num_defs;
    const unsigned num_ops = InsOpcode(ins).num_operands;
    for (unsigned i = num_defs; i < num_ops; ++i) {
      if (InsOperand(ins, i) == reg_src) InsOperand(ins, i) = reg_dst;
    }
    for (unsigned i = 0; i < num_defs; ++i) {
      if (InsOperand(ins, i) == reg_src) return;
    }
    ins = BblInsList::Next(ins);
  }
}

}  // namespace

int FunSeparateLocalRegUsage(Fun fun) {
  int count = 0;
  for (Bbl bbl : FunBblIter(fun)) {
    for (Ins ins : BblInsIter(bbl)) {
      const unsigned num_defs = InsOpcode(ins).num_defs;
      for (unsigned i = 0; i < num_defs; ++i) {
        Reg reg = Reg(InsOperand(ins, i));
        if (RegDefIns(reg) == ins || RegHasFlag(reg, REG_FLAG::GLOBAL) ||
            (RegHasFlag(reg, REG_FLAG::TWO_ADDRESS) &&
             InsOpcode(ins).num_operands >= 2 &&
             InsOperand(ins, 0) == InsOperand(ins, 1)) ||
            RegCpuReg(reg) != HandleInvalid) {
          continue;
        }
        auto purpose = MaybeSkipCountPrefix(StrData(Name(reg)));
        const Reg new_reg = FunGetScratchReg(fun, RegKind(reg), purpose, false);
        if (RegHasFlag(reg, REG_FLAG::TWO_ADDRESS)) {
          RegFlags(new_reg) |= +REG_FLAG::TWO_ADDRESS;
        }
        InsOperand(ins, i) = new_reg;
        RenameRegRange(BblInsList::Next(ins), reg, new_reg);
        count += 1;
      }
    }
  }
  return count;
}

DK_LAC_COUNTS FunGlobalRegStats(Fun fun, const DK_MAP& rk_map) {
  DK_LAC_COUNTS out;
  for (Reg reg : FunRegIter(fun)) {
    if (!RegCpuReg(reg).isnull() || !RegHasFlag(reg, REG_FLAG::GLOBAL)) {
      continue;
    }
    const unsigned kind = rk_map[+RegKind(reg)];
    ASSERT(kind != 0, "");
    if (RegHasFlag(reg, REG_FLAG::LAC))
      ++out.lac[kind];
    else
      ++out.not_lac[kind];
  }
  return out;
}

}  // namespace cwerg::base
